### PR TITLE
Fix wrapped token logic

### PR DESCRIPTION
**Pull request by amonsosanz**
### What does this PR does?

This pull request fixes several bugs and edge cases. It requires an in-depth test session bridging ETH, each of the ERC-20 tokens in the list (they implement different permits) and some custom tokens both native to L1 and L2, back and forth.

Some of the fixes:
1. Incorrectly inferring when a token was native or wrapped and therefore computing incorrect token addresses.
1. Fixing the lack of support for different tokens that share the same address, one in L1 and the other in L2.
1. Wrongly selecting the addresses of L2 native tokens when initiating a bridge. The error happened when checking the `allowance` and doing the approval/permit (@invocamanman bug).
1. Enable permit for DAI
1. Increase claim gas limit to support new tokens deployment on L2
1. Fix MATIC token name

Here is some guidance to test the main relevant aspects:

### How to test 1 and 3?

* Setup the env to work on https://rpc.public.zkevm-test.net (mango)
* Select the zkEVM network
* Add this custom token `0x2A7B29B373CDe8459Dd66e78b56b9c6c1d79Ad96`
* Mint some tokens using the [mint](https://explorer.public.zkevm-test.net/address/0x2A7B29B373CDe8459Dd66e78b56b9c6c1d79Ad96/write-contract#address-tabs) method (6)
* Bridge some tokens to L1
* Bridge some tokens back to L2

### How to test 2?

* Setup the env to work on https://rpc.public.zkevm-test.net (mango)
* Add this token in L1 `0x0B5f1c0E6E3BFEc85e044dB01efc3a4B03d67c69`
* Add this token in L2 `0x0B5f1c0E6E3BFEc85e044dB01efc3a4B03d67c69` (it's the same address but it's a different token)
* Mint some tokens in L2 using the [mint](https://explorer.public.zkevm-test.net/address/0x0B5f1c0E6E3BFEc85e044dB01efc3a4B03d67c69/write-contract#address-tabs) method (6)
* Bridge some tokens to L1
* Bridge some tokens back to L2
* Make sure that both Activity and TX details show the right token.
* Remove the tokens and make sure the right token is removed.
* Make sure that both Activity and TX details show the right info with the tokens removed.

### How to test 4?

* Setup the env to work on https://rpc.public.zkevm-test.net (mango)
* Mint some DAI tokens using the [mint](https://goerli.etherscan.io/address/0xD7E55eB808693D5Ff81a3391c59886C7E0449f35#writeContract) method (4)
* Bridge some DAI to L2
* Bridge some DAI back to L1

### CODE REVIEW

File: src/contexts/tokens.context.tsx [hunk 1]
```diff
@@ -32,8 +32,8 @@ interface ComputeWrappedTokenAddressParams {
 }
 
 interface GetNativeTokenInfoParams {
+  address: string;
```
Comment (elias-garcia) [inline src/contexts/tokens.context.tsx:36]: I would still pass the domain interface as we are doing in the bridge instead of just the address

File: src/contexts/tokens.context.tsx [hunk 2]
```diff
@@ -158,78 +158,90 @@ const TokensProvider: FC<PropsWithChildren> = (props) => {
...
+    async ({ address }: GetTokenFromAddressParams): Promise<Token> => {
+      if (!env) {
+        throw Error("The env is not ready");
+      }
+
+      if (!isAsyncTaskDataAvailable(connectedProvider)) {
+        throw Error("The connectedProvider is not ready");
+      }
+
+      const chain = env.chains.find((chain) => chain.chainId === connectedProvider.data.chainId);
```
Comment (elias-garcia) [inline src/contexts/tokens.context.tsx]: Didn't we agree on use the `provider` that was currently selected in the form? The `connectedProvider` could be different than the one selected in the form.

File: src/contexts/tokens.context.tsx (hunk 1 above)
Comment (amonsosanz) [reply src/contexts/tokens.context.tsx:36]: We can't because [we still don't have the token built](https://github.com/0xPolygonHermez/zkevm-bridge-ui/pull/241/files#diff-9c5cb3f673be0ba52c8570a1ccd23fcfab6bfc5ee91735b11a42d160c140f6e0R197).

File: src/contexts/tokens.context.tsx (hunk 2 above)
Comment (amonsosanz) [reply src/contexts/tokens.context.tsx]: mmmm.... what I understood is that we were going to use the provider of the selected network in metamask, not in the form.

Comment (elias-garcia) [reply src/contexts/tokens.context.tsx]: Take into account that when you are in the form and you search for a token, you'd like to search for it in the network currently selected in the form. If not, if the connected provider is connected to the other network and the token doesn't exist in the network currently selected in the form you'll see an error saying that the token doesn't exist when you expect it to exist.

Comment (amonsosanz) [reply src/contexts/tokens.context.tsx]: Ok, I will change this, but my reasoning was the same, but for the globally selected network: When you search for a token, you'd like to search for it in the network currently selected in MetaMask.

Having 2 selected networks is a clear design flaw in this application. Some users might expect the behavior you describe, but I'm sure some users would expect the behavior I describe.

Comment (elias-garcia) [reply src/contexts/tokens.context.tsx]: Well, I don't think so. Usually, when you are in a dapp playing around usually you forget about the network selected in MetaMask very easily. And if the UI has one chain selected, you are going to think that you are connected to that network, so you expect to search the token in that chain. As an example, this is what Jesús was doing the other that.

Comment (amonsosanz) [reply src/contexts/tokens.context.tsx]: Done

**Comment by elias-garcia** [comment]
Regarding this topic, I've seen this error reported by some users:

```
{"reason":null,"code":"CALL_EXCEPTION","method":"precalculatedWrapperAddress(uint32,address,string,string,uint8)","data":"0x","errorArgs":null,"errorName":null,"errorSignature":null,"address":"0x9d98deabc42dd696deb9e40b4f1cab7ddbf55988","args":[0,"0x0000000000000000000000000000000000000000","Ether","ETH",18],"transaction":{"data":"0xaaa13cc20000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000e000000000000000000000000000000000000000000000000000000000000000120000000000000000000000000000000000000000000000000000000000000005457468657200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000034554480000000000000000000000000000000000000000000000000000000000","to":"0x9D98DeAbC42dd696Deb9e40b4f1CAB7dDBF55988"}}
```

Looks like we are trying to call the `precalculatedWrapperAddress` function with the `ETH` token and it's producing a `CALL_EXCEPTION`. I think that it's worth to check this bug in this PR as well.

**Comment by elias-garcia** [comment]
## Test 1 & 3

I was able to bridge ALEXtoken from L2 to L1, but when I try to bridge it back to L2 all the transactions are failing, like [this one](https://explorer.public.zkevm-test.net/tx/0x4764fda1ea69019338d8f694c7e11c4d0ea8f11cc739a52ec49b17c5cfe70c28).

## Test 4

I was able to bridge DAI from L1 to L2, but when trying to bridge them back to L1 I'm getting an `execution reverted: PolygonZkEVMBridge::_verifyLeaf: Invalid smt proof` error:

```
cannot estimate gas; transaction may fail or may require manual gas limit [ See: https://links.ethers.org/v5-errors-UNPREDICTABLE_GAS_LIMIT ] (reason="execution reverted: PolygonZkEVMBridge::_verifyLeaf: Invalid smt proof", method="estimateGas", transaction={"from":"0x45279cB343D8244b0D2F2c3274bC37d8F0F2C303","to":"0xe2E52C2D0D64209b8DD1854371A4C673c13448f0","data":"0x7b6323c100000000000000000000000000000000000000000000000000000000000001400000000000000000000000000000000000000000000000000000000000007ed4145c0da38314d95f33418a10f398d755207c1201ea850b01ff41f70f2a9b88c712d3093685179d107ef03c86b3094cf9fa20174149e747f4fd342c13c1c7ff030000000000000000000000000000000000000000000000000000000000000000000000000000000000000000ffe9a0f61f55b6fb8b88aecbde698a5ef1495e5e000000000000000000000000000000000000000000000000000000000000000000000000000000000000000045279cb343d8244b0d2f2c3274bc37d8f0f2c3030000000000000000000000000000000000000000000000056bc75e2d63100000000000000000000000000000000000000000000000000000000000000000056000000000000000000000000000000000000000000000000000000000000000209a2790d7d8477711b357aae190085880a85c6c9a2fd46c1a9debab3cb60a8a30ba2e49e0bfe11dcef652e08fb167316f75b952c581dc07b13b555c68e08d2df0a9d2c3f4cfc82ddbc02ef15ce5657bb6ecbfdd3c7c0c4b161db5b6c269a9c41e3f6fdfd9c60f138bb89553eec8a358a8a26655d78822ec2f61dba6db7a6a0f34545a1438f5ea4de5a203ed191e9aa401c70155354e28d80bde66dfa49910e3fec13a754554210c0e7db8ac10bba90c2760ee9cdc5ff7f917dc547622956b539dbad917ea1f7204c486f840fafa19427fc797b9f25f9245d11b5df5bebb55ff718d77623ea0876533a7e7c607282819a308b71982cc070dfd49fa918a98e674c19867cc5f7f196b93bae1e27e6320742445d290f2263827498b54fec539f756af8cbc916b0b479ab30533a84155685881a9ae3944b481dfa8b12ef1ce7f0312602958ecfe92e3e740c5717d357f6827c6893a4216f380f4b36f3a632c2ac72266d16dcc93c2bda678a37deac5d73dd76dc9e444522b2ca30a84a594e5581f1d1935354f4fd093bf7dac967e0bfc28a99302247b0ddcf39bd40a8ad31061bd3a2c1b6efbe86aeec7272ab75ccd87944b09631eaf4bb274c806c475142ee8b4187dfa8c85824402e96d48bd2946dc13049861c67712e44dd4824c536f0da2b0789cda7bce9f4e8618b6bd2f4132ce798cdc7a60e7e1460a7299e3c6342a579626d22733e50f526ec2fa19a22b31e8ed50f23cd1fdf94c9154ed3a7609a2f1ff981fe1d3b5c807b281e4683cc6d6315cf95b9ade8641defcb32372f1c126e398ef7a5a2dce0a8a7f68bb74560f8f71837c2c2ebbcbf7fffb42ae1896f13f7c7479a0b46a28b6f55540f89444f63de0378e3d121be09e06cc9ded1c20e65876d36aa0c65e9645644786b620e2dd2ad648ddfcbf4a7e5b1a3a4ecfe7f64667a3f0b7e2f4418588ed35a2458cffeb39b93d26f18d2ab13bdce6aee58e7b99359ec2dfd95a9c16dc00d6ef18b7933a6f8dc65ccb55667138776f7dea101070dc8796e3774df84f40ae0c8229d0d6069e5c8f39a7c299677a09d367fc7b05e3bc380ee652cdc72595f74c7b1043d0e1ffbab734648c838dfb0527d971b602bc216c9619ef0abf5ac974a1ed57f4050aa510dd9c74f508277b39d7973bb2dfccc5eeb0618db8cd74046ff337f0a7bf2c8e03e10f642c1886798d71806ab1e888d9e5ee87d0838c5655cb21c6cb83313b5a631175dff4963772cce9108188b34ac87c81c41e662ee4dd2dd7b2bc707961b1e646c4047669dcb6584f0d8d770daf5d7e7deb2e388ab20e2573d171a88108e79d820e98f26c0b84aa8b2f4aa4968dbb818ea32293237c50ba75ee485f4c22adf2f741400bdf8d6a9cc7df7ecae576221665d7358448818bb4ae4562849e949e17ac16e0be16688e156b5cf15e098c627c0056a90000000000000000000000000000000000000000000000000000000000000000","accessList":null}, error={"code":-32603,"message":"execution reverted: PolygonZkEVMBridge::_verifyLeaf: Invalid smt proof","data":{"originalError":{"code":3,"data":"0x08c379a000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000032506f6c79676f6e5a6b45564d4272696467653a3a5f7665726966794c6561663a20496e76616c696420736d742070726f6f660000000000000000000000000000","message":"execution reverted: PolygonZkEVMBridge::_verifyLeaf: Invalid smt proof"}}}, code=UNPREDICTABLE_GAS_LIMIT, version=providers/5.6.8)
```

**Comment by elias-garcia** [comment]
> ## Test 1 & 3
> I was able to bridge ALEXtoken from L2 to L1, but when I try to bridge it back to L2 all the transactions are failing, like [this one](https://explorer.public.zkevm-test.net/tx/0x4764fda1ea69019338d8f694c7e11c4d0ea8f11cc739a52ec49b17c5cfe70c28).
> 
> ## Test 4
> I was able to bridge DAI from L1 to L2, but when trying to bridge them back to L1 I'm getting an `execution reverted: PolygonZkEVMBridge::_verifyLeaf: Invalid smt proof` error:
> 
> ```
> cannot estimate gas; transaction may fail or may require manual gas limit [ See: https://links.ethers.org/v5-errors-UNPREDICTABLE_GAS_LIMIT ] (reason="execution reverted: PolygonZkEVMBridge::_verifyLeaf: Invalid smt proof", method="estimateGas", transaction={"from":"0x45279cB343D8244b0D2F2c3274bC37d8F0F2C303","to":"0xe2E52C2D0D64209b8DD1854371A4C673c13448f0","data":"0x7b6323c100000000000000000000000000000000000000000000000000000000000001400000000000000000000000000000000000000000000000000000000000007ed4145c0da38314d95f33418a10f398d755207c1201ea850b01ff41f70f2a9b88c712d3093685179d107ef03c86b3094cf9fa20174149e747f4fd342c13c1c7ff030000000000000000000000000000000000000000000000000000000000000000000000000000000000000000ffe9a0f61f55b6fb8b88aecbde698a5ef1495e5e000000000000000000000000000000000000000000000000000000000000000000000000000000000000000045279cb343d8244b0d2f2c3274bc37d8f0f2c3030000000000000000000000000000000000000000000000056bc75e2d63100000000000000000000000000000000000000000000000000000000000000000056000000000000000000000000000000000000000000000000000000000000000209a2790d7d8477711b357aae190085880a85c6c9a2fd46c1a9debab3cb60a8a30ba2e49e0bfe11dcef652e08fb167316f75b952c581dc07b13b555c68e08d2df0a9d2c3f4cfc82ddbc02ef15ce5657bb6ecbfdd3c7c0c4b161db5b6c269a9c41e3f6fdfd9c60f138bb89553eec8a358a8a26655d78822ec2f61dba6db7a6a0f34545a1438f5ea4de5a203ed191e9aa401c70155354e28d80bde66dfa49910e3fec13a754554210c0e7db8ac10bba90c2760ee9cdc5ff7f917dc547622956b539dbad917ea1f7204c486f840fafa19427fc797b9f25f9245d11b5df5bebb55ff718d77623ea0876533a7e7c607282819a308b71982cc070dfd49fa918a98e674c19867cc5f7f196b93bae1e27e6320742445d290f2263827498b54fec539f756af8cbc916b0b479ab30533a84155685881a9ae3944b481dfa8b12ef1ce7f0312602958ecfe92e3e740c5717d357f6827c6893a4216f380f4b36f3a632c2ac72266d16dcc93c2bda678a37deac5d73dd76dc9e444522b2ca30a84a594e5581f1d1935354f4fd093bf7dac967e0bfc28a99302247b0ddcf39bd40a8ad31061bd3a2c1b6efbe86aeec7272ab75ccd87944b09631eaf4bb274c806c475142ee8b4187dfa8c85824402e96d48bd2946dc13049861c67712e44dd4824c536f0da2b0789cda7bce9f4e8618b6bd2f4132ce798cdc7a60e7e1460a7299e3c6342a579626d22733e50f526ec2fa19a22b31e8ed50f23cd1fdf94c9154ed3a7609a2f1ff981fe1d3b5c807b281e4683cc6d6315cf95b9ade8641defcb32372f1c126e398ef7a5a2dce0a8a7f68bb74560f8f71837c2c2ebbcbf7fffb42ae1896f13f7c7479a0b46a28b6f55540f89444f63de0378e3d121be09e06cc9ded1c20e65876d36aa0c65e9645644786b620e2dd2ad648ddfcbf4a7e5b1a3a4ecfe7f64667a3f0b7e2f4418588ed35a2458cffeb39b93d26f18d2ab13bdce6aee58e7b99359ec2dfd95a9c16dc00d6ef18b7933a6f8dc65ccb55667138776f7dea101070dc8796e3774df84f40ae0c8229d0d6069e5c8f39a7c299677a09d367fc7b05e3bc380ee652cdc72595f74c7b1043d0e1ffbab734648c838dfb0527d971b602bc216c9619ef0abf5ac974a1ed57f4050aa510dd9c74f508277b39d7973bb2dfccc5eeb0618db8cd74046ff337f0a7bf2c8e03e10f642c1886798d71806ab1e888d9e5ee87d0838c5655cb21c6cb83313b5a631175dff4963772cce9108188b34ac87c81c41e662ee4dd2dd7b2bc707961b1e646c4047669dcb6584f0d8d770daf5d7e7deb2e388ab20e2573d171a88108e79d820e98f26c0b84aa8b2f4aa4968dbb818ea32293237c50ba75ee485f4c22adf2f741400bdf8d6a9cc7df7ecae576221665d7358448818bb4ae4562849e949e17ac16e0be16688e156b5cf15e098c627c0056a90000000000000000000000000000000000000000000000000000000000000000","accessList":null}, error={"code":-32603,"message":"execution reverted: PolygonZkEVMBridge::_verifyLeaf: Invalid smt proof","data":{"originalError":{"code":3,"data":"0x08c379a000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000032506f6c79676f6e5a6b45564d4272696467653a3a5f7665726966794c6561663a20496e76616c696420736d742070726f6f660000000000000000000000000000","message":"execution reverted: PolygonZkEVMBridge::_verifyLeaf: Invalid smt proof"}}}, code=UNPREDICTABLE_GAS_LIMIT, version=providers/5.6.8)
> ```

I've been checking this errors with Jesús, and the origin of both errors is that the `originTokenAddress` is not being set correctly when calling the `claimAsset` method.

EDIT: The problem is in the `selectTokenAddress` function, which is selecting the address based on the chain on the `claim` function, when here the `originTokenAddress` is always the same (it doesn't change based on the chain).

**Comment by elias-garcia** [comment]
I also see that now, the tokens in the list (not custom tokens) only show their L1 address but not their L2 address. Is this expected and if true, why?

Captura de pantalla 2023-01-04 a las 11 40 29

**Comment by elias-garcia** [comment]
> I also see that now, the tokens in the list (not custom tokens) only show their L1 address but not their L2 address. Is this expected and if true, why?
> 
> Captura de pantalla 2023-01-04 a las 11 40 29

Looks like this could be a network issue as I was having a lot of problems due to rate limit, I can now see the L2 address correctly as before.

File: src/contexts/bridge.context.tsx [hunk 3]
```diff
@@ -865,7 +852,7 @@ const BridgeProvider: FC<PropsWithChildren> = (props) => {
             destinationAddress,
             amount,
             metadata,
-            isL2Claim ? { gasLimit: 500000, gasPrice: 0 } : {}
+            isL2Claim ? { gasLimit: 1500000, gasPrice: 0 } : {}
```
Comment (elias-garcia) [inline src/contexts/bridge.context.tsx:855]: 1M and a half gas for a L2 claim? Is this correct? I've done some L2 claims and and the txs spent 5% of this gas

File: src/contexts/tokens.context.tsx [hunk 4]
```diff
@@ -158,104 +154,107 @@ const TokensProvider: FC<PropsWithChildren> = (props) => {
...
-          return await getTokenFromAddress({ address: tokenAddress, chain }).catch(() => {
-            throw new Error(
-              `The token with the address "${tokenAddress}" could not be found either in the list of supported Tokens or in the blockchain with network id "${originNetwork}"`
-            );
-          });
-        } else {
+        return getTokenFromAddress({ address: tokenOriginAddress, chain }).catch(() => {
           throw new Error(
-            `The token with the address "${tokenAddress}" could not be found in the list of supported Tokens and the provided network id "${originNetwork}" is not supported`
+            `The token with the address "${tokenOriginAddress}" could not be found either in the list of supported Tokens or in the blockchain with network id "${originNetwork}"`
```
Comment (elias-garcia) [inline src/contexts/tokens.context.tsx]: The network id is something special from the zkEVM, I would use the Chain Id here, or even the Chain Name

File: src/contexts/bridge.context.tsx (hunk 3 above)
Comment (amonsosanz) [reply src/contexts/bridge.context.tsx:855]: Yep, this is required only when a contract is a new contract needs to be deployed. @invocamanman asked me to do this change.

Comment (elias-garcia) [reply src/contexts/bridge.context.tsx:855]: Ahh okay! Yeah, I remember that Jesús told me that he was going to change the implementation of new contracts and that they would use far more gas than now.

File: src/contexts/tokens.context.tsx (hunk 4 above)
Comment (amonsosanz) [reply src/contexts/tokens.context.tsx]: Done

**Comment by amonsosanz** [comment]
> Looks like we are trying to call the `precalculatedWrapperAddress` function with the `ETH` token and it's producing a `CALL_EXCEPTION`. I think that it's worth to check this bug in this PR as well.

Fixed.

**Review by elias-garcia — APPROVED** [review]
